runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/JavaReadViaImpulseTest.java

```diff
@@ -89,14 +96,50 @@ public void testSplitSourceFn() {
   public void testReadFromSourceFn() {
     BoundedSource<Long> source = CountingSource.upTo(10L);
     PCollection<BoundedSource<Long>> sourcePC =
-        (PCollection)
-            p.apply(Create.of(source).withCoder(SerializableCoder.of((Class) BoundedSource.class)));
-    PCollection<Long> elems = sourcePC.apply(ParDo.of(new ReadFromBoundedSourceFn<>()));
+          p.apply(Create.of(source)
+              .withCoder(new JavaReadViaImpulse.BoundedSourceCoder<>()));
+    PCollection<Long> elems = sourcePC.apply(ParDo.of(new ReadFromBoundedSourceFn<>()))
+        .setCoder(VarLongCoder.of());
 
     PAssert.that(elems).containsInAnyOrder(0L, 9L, 8L, 1L, 2L, 7L, 6L, 3L, 4L, 5L);
     p.run();
   }
 
+  @Test
+  @Category(NeedsRunner.class)
+  public void testReadToImpulseOverride() {
+    BoundedSource<Long> source = CountingSource.upTo(10L);
+    // Use an explicit read transform to ensure the override is exercised.
+    PCollection<Long> input = p.apply(Read.from(source));
+    PAssert.that(input).containsInAnyOrder(0L, 1L, 2L, 3L, 4L, 5L, 6L, 7L, 8L, 9L);
+    p.replaceAll(Collections.singletonList(JavaReadViaImpulse.boundedOverride()));
+    p.traverseTopologically(new Pipeline.PipelineVisitor() {
```

Comment:
extend PipelineVisitor.Defaults, and get rid of `enterPipeline`, `visitValue`, and `leave*`

Reply:
This is actually already in master and was just one of the dependent changes.